plugins/package-managers/bazel/src/main/kotlin/BazelModel.kt

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.packagemanagers.bazel
+
+import java.io.File
+import java.util.Base64
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.BazelModuleRegistryClient
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleMetadata
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleSourceInfo
+import org.ossreviewtoolkit.model.Hash
+import org.ossreviewtoolkit.model.HashAlgorithm
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.VcsType
+
+private val json = Json { ignoreUnknownKeys = true }
+
+/**
+ * The data model for MODULE.bazel.lock.
+ */
+@Serializable
+internal data class LockFile(
+    val lockFileVersion: Int,
+    val flags: Flags,
+    val moduleDepGraph: Map<String, DepGraphEntry>
+) {
+    // TODO Support multiple registries.
+    fun registryUrl(): String = flags.cmdRegistries.getOrElse(0) { BazelModuleRegistryClient.DEFAULT_URL }
+}
+
+@Serializable
+internal data class Flags(
+    val cmdRegistries: List<String>
+)
+
+@Serializable
+internal data class DepGraphEntry(
+    val name: String,
+    val version: String,
+    val key: String,
+    val repoName: String,
+    val deps: Map<String, String>,
+    val repoSpec: RepoSpec? = null
+)
+
+@Serializable
+internal data class RepoSpec(
+    val bzlFile: String,
+    val ruleClassName: String,
+    val attributes: RepoAttributes
+)
+
+@Serializable data class RepoAttributes(
+    val urls: List<String>,
+    val integrity: String,
+    @SerialName("strip_prefix")
+    val stripPrefix: String,
+    @SerialName("remote_patches")
+    val remotePatches: Map<String, String>,
+    @SerialName("remote_patch_strip")
+    val remotePatchStrip: Int
+)
+
+internal fun parseLockfile(lockFile: File): LockFile = json.decodeFromString<LockFile>(lockFile.readText())
+
+/**
+ * The data model for the output of "bazel mod graph --output json".
+ */
+@Serializable
+internal data class BzlmodGraphNode(
+    val key: String,
+    val name: String? = null,
+    val version: String? = null,
+    val dependencies: List<BzlmodGraphNode> = emptyList()
+) {
+    fun name(): String {
+        if (name != null) {
+            return name
+        }
+
+        val parts = key.split('@')
+        return parts.getOrElse(0) { "" }
+    }
+
+    fun version(): String {
+        if (version != null) {
+            return version
+        }
+
+        val parts = key.split('@')
+        return parts.getOrElse(1) { "" }
+    }
+}
+
+fun toVcsInfo(metadata: ModuleMetadata): VcsInfo {
+    if (metadata.repository.isNullOrEmpty()) {
+        return VcsInfo.EMPTY
+    }
+
+    val repo = metadata.repository!![0]
```

Comment:
I don't know how to avoid this.

Reply:
Since `repository` is a nullable property (and also the list could be empty), I'd use

    val repo = metadata.repository?.firstOrNull()

and handle the `null`-case of `repo` accordingly in the code.